### PR TITLE
Fix demos to use jax that require grads

### PR DESCRIPTION
**Pull request by austingmhuang**
Doesn't build due to the fact that [this PR](https://github.com/PennyLaneAI/pennylane/pull/6096) is not yet merged. JSON files need to be updated as well.

**Summary:**
Fixes remaining demos that checked for requires_grad. Won't build for now due to missing a fix in [this PR](https://github.com/PennyLaneAI/pennylane/pull/6324)

[[sc-69776](https://app.shortcut.com/xanaduai/story/69776)] [[sc-69778](https://app.shortcut.com/xanaduai/story/69778)]

### CODE REVIEW

File: demonstrations/tutorial_mol_geo_opt.py [hunk 1]
```diff
@@ -191,12 +194,12 @@ def H(x):
 # The ``hf`` array is used by the :class:`~.pennylane.BasisState` operation to initialize
 # the qubit register. Then, the :class:`~.pennylane.DoubleExcitation` operations are applied
 # First, we define the quantum device used to compute the expectation value.
-# In this example, we use the ``lightning.qubit`` simulator:
+# In this example, we use the ``default.qubit`` simulator:
 num_wires = 6
-dev = qml.device("lightning.qubit", wires=num_wires)
+dev = qml.device("default.qubit", wires=num_wires)
```
Comment (austingmhuang) [inline demonstrations/tutorial_mol_geo_opt.py:199]: For some reason lightning.qubit doesn't work with JAX here. Not sure why but I had to change to using default to get it working.

File: demonstrations/tutorial_resource_estimation.py [hunk 2]
```diff
@@ -65,12 +65,16 @@
 `6-31g basis set <https://en.wikipedia.org/wiki/Basis_set_(chemistry)>`_ as an example.
 """
 import pennylane as qml
-from pennylane import numpy as np
+import numpy as np
+from jax import numpy as jnp
+import jax
```
Comment (soranjh) [inline demonstrations/tutorial_resource_estimation.py:70]: Let's just use vanilla numpy in this demo.

File: demonstrations/tutorial_mapping.py [hunk 3]
```diff
@@ -202,11 +202,11 @@
...
 
 from pennylane import qchem
-from pennylane import numpy as np
+from jax import numpy as jnp
 
 symbols  = ['H', 'H']
-geometry = np.array([[0.0, 0.0, -0.69434785],
-                     [0.0, 0.0,  0.69434785]], requires_grad = False)
+geometry = jnp.array([[0.0, 0.0, -0.69434785],
+                     [0.0, 0.0,  0.69434785]])
```
Comment (soranjh) [inline demonstrations/tutorial_mapping.py:209]: This demo is already updated in another [PR](https://github.com/PennyLaneAI/qml/pull/1211), correct?

File: demonstrations/tutorial_fermionic_operators.py [hunk 4]
```diff
@@ -108,7 +108,7 @@
 # The matrix representation of the qubit Hamiltonian in the computational basis can be diagonalized
 # to get its eigenpairs.
 
-from pennylane import numpy as np
+import numpy as np
```
Comment (soranjh) [inline demonstrations/tutorial_fermionic_operators.py:111]: Can we update this demo in this [PR](https://github.com/PennyLaneAI/qml/pull/1211) as well?

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @austingmhuang. befor doing a full review, could you please make sure that this PR targets dev instead of master? Also, if there is any demo that does not require features in dev, their upgrade should be done in this [PR](https://github.com/PennyLaneAI/qml/pull/1211).

File: demonstrations/tutorial_mapping.py (hunk 3 above)
Comment (austingmhuang) [reply demonstrations/tutorial_mapping.py:209]: No, it's not updated in that PR. Well that PR had a one line change in this demo by accident.

Comment (austingmhuang) [reply demonstrations/tutorial_mapping.py:209]: This demo doesn't work without the requires_grad changes because it calls fermionic_hamiltonian directly

File: demonstrations/tutorial_fermionic_operators.py (hunk 4 above)
Comment (austingmhuang) [reply demonstrations/tutorial_fermionic_operators.py:111]: No, it cannot be done. This demo calls `electronic_integrals()` directly, which checks for requires_grad, leading to an error.

File: demonstrations/tutorial_differentiable_HF.py [hunk 5]
```diff
@@ -228,13 +230,13 @@
 # plane.
 
 n = 30 # number of grid points along each axis
-
+qml.qchem.hf_energy(mol)()
```
Comment (KetpuntoG) [inline demonstrations/tutorial_differentiable_HF.py:233]: What is this line doing?

File: demonstrations/tutorial_differentiable_HF.py [hunk 6]
```diff
@@ -313,44 +317,41 @@ def circuit(*args):
...
+args = [circuit_param, geometry, coeff, alpha]
 
-    args = [circuit_param, geometry, alpha, coeff]
+for n in range(36):
+    args = [circuit_param, geometry, coeff, alpha]
     mol = qml.qchem.Molecule(symbols, geometry, alpha=alpha, coeff=coeff)
 
     # gradient for circuit parameters
-    g_param = grad(energy(mol), argnum=0)(*args)
+    g_param = jax.grad(energy(), argnums=[0, 1, 2, 3])(*args)[0]
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply demonstrations/tutorial_differentiable_HF.py:344]: Yes 👍

[the remark this answers is not in the document]
Comment (austingmhuang) [reply demonstrations/tutorial_differentiable_HF.py:344]: We were doing that beforehand too (geom, alpha, coeff, and circuit_param were all labelled with `requires_grad=True`. We just update the UI to use jax UI instead.

File: demonstrations/tutorial_differentiable_HF.py (hunk 5 above)
Comment (austingmhuang) [reply demonstrations/tutorial_differentiable_HF.py:233]: It is a bit of a hack to reset the information in `mol` to use numpy arrays instead of jax arrays. When we did `jax.grad(qml.qchem.hf_energy(mol), argnums=0)(geometry, mol.coeff, mol.alpha)`, internally some of the information of the molecule automatically uses jax arrays. This is great, but not for this graphing part, where we want to stick to using numpy.

File: demonstrations/tutorial_differentiable_HF.py [hunk 7]
```diff
@@ -101,16 +101,18 @@
...
 
 symbols = ["H", "H"]
 # optimized geometry at the Hartree-Fock level
-geometry = np.array([[-0.672943567415407, 0.0, 0.0],
-                     [ 0.672943567415407, 0.0, 0.0]], requires_grad=True)
+geometry = jnp.array([[-0.672943567415407, 0.0, 0.0],
+                     [ 0.672943567415407, 0.0, 0.0]])
 
 ##############################################################################
 # The use of ``requires_grad=True`` specifies that the nuclear coordinates are differentiable
```
Comment (austingmhuang) [inline demonstrations/tutorial_differentiable_HF.py:118]: We should delete this line.

File: demonstrations/tutorial_mol_geo_opt.py [hunk 8]
```diff
@@ -312,44 +310,39 @@ def grad_x(params, x):
...
+    theta = theta - 0.8 * g_param
 
-    energy.append(cost(theta, x))
-    bond_length.append(np.linalg.norm(x[0:3] - x[3:6]) * bohr_angs)
+    # gradient for coordinates
+    value, _ = jax.value_and_grad(cost, argnums=1)(theta, x)
+    grad = grad_x(theta, x)
+    x = x - 0.8 * grad
+    energies.append(value)
+    bond_length.append(jnp.linalg.norm(x[0] - x[1]) * bohr_angs)
```
[the remark this answers is not in the document]
Comment (austingmhuang) [reply demonstrations/tutorial_mol_geo_opt.py:332]: `x` is still the nuclear gradients. Just now we are manually updating its value instead of letting opt_x.step() take care of that.

**Review by KetpuntoG — APPROVED** [review]
Look good! please update the metadata before merging
